### PR TITLE
Cbrune/bfd549 enrollmentdata

### PR DESCRIPTION
**Pull request by cbrunefearless**
### Change Details
Added enrollment data schema to keep track by year and month.

### Acceptance Validation

Review schema and also the code and Integration tests

### Feedback Requested

Integration tests test the features being changed.
### External References

- [BFD-549](https://jira.cms.gov/browse/BFD-549)

### Security Implications

- [ ] new software dependencies
N/A

- [ ] altered security controls

- [ ] new data stored or transmitted

- [ ] security checklist is completed for this change

- [ ] requires more information or team discussion to evaluate security implications

### CODE REVIEW

File: apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java
Comment (karlmdavis) [inline apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java:999]: Not sure we want this to be lazy-loaded. Any thoughts?

File: apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java
Comment (karlmdavis) [inline apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java:1974]: Can the pre-existing `selectJavaFieldType(...)` and `selectJavaPropertyType(...)` methods be refactored to call these under the hood? If so, we should definitely do that, to cut down on maintenance reqs.

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V27__Create_Enrollmenttable.sql
Comment (karlmdavis) [inline apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V27__Create_Enrollmenttable.sql:137]: Are there any `Beneficiaries` records where `beneEnrollmentReferenceYear` is `NULL`? If so, what does that indicate about the bene record and how should we handle it here?

File: apps/bfd-pipeline/bfd-pipeline-rif-load/src/main/java/gov/cms/bfd/pipeline/rif/load/RifLoader.java
Comment (karlmdavis) [inline apps/bfd-pipeline/bfd-pipeline-rif-load/src/main/java/gov/cms/bfd/pipeline/rif/load/RifLoader.java:849]: I think this will overwrite **all** enrollment records, instead of just the current year's won't it? If you want to avoid that, you need to merge the new/current records into the existing list, avoiding duplicates (which would likely cause PK errors).

More importantly, if this wasn't caught by your tests, then I think you need more ITs to ensure that all of this is actually accomplishing the thing we need it to.

File: apps/bfd-pipeline/bfd-pipeline-rif-load/src/test/java/gov/cms/bfd/pipeline/rif/load/RifLoaderIT.java
Comment (karlmdavis) [inline apps/bfd-pipeline/bfd-pipeline-rif-load/src/test/java/gov/cms/bfd/pipeline/rif/load/RifLoaderIT.java:237]: As mentioned above, I think you'll want to have several test cases here, covering the following scenarios:
1. If we insert a bene, do the resulting DB records have only the correct months' data populated, or does it end up with a lot of weird null-ish entries?
2. If we update a bene with the current year's data and just more months filled in, does the resulting `Enrollments` data look as expected?
3. If we update a bene with the current year's data and *less* months filled in (shouldn't happen but the schema allows it so we should ahndle it just in case), do the resulting DB records look as expected?
4. If we update a bene with a new year's data, is it appended to the previous years' data, or does it overwrite it?

**Review by karlmdavis — CHANGES_REQUESTED** [review]
This is really impressive work in a very gnarly part of the codebase, Chris! Bravo!

However, I'm not sure -- and apologies if I'm wrong -- but I think this still has a major bug in it. In addition to fixing that, I'd suggest adding test coverage (as detailed in one of the comments). I also suggested a couple of other minor tweaks.

File: apps/bfd-pipeline/bfd-pipeline-rif-load/src/main/java/gov/cms/bfd/pipeline/rif/load/RifLoader.java
Comment (cbrunefearless) [reply apps/bfd-pipeline/bfd-pipeline-rif-load/src/main/java/gov/cms/bfd/pipeline/rif/load/RifLoader.java:849]: Yea Changed the logic and adding more IT's

File: apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java
Comment (cbrunefearless) [reply apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java:999]: Eager loaded

File: apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java
Comment (cbrunefearless) [reply apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java:1974]: yes done..

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V27__Create_Enrollmenttable.sql
Comment (cbrunefearless) [reply apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V27__Create_Enrollmenttable.sql:137]: what do u propose we should do if its null?

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V27__Create_Enrollmenttable.sql
Comment (karlmdavis) [inline apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V27__Create_Enrollmenttable.sql:10]: Why this column type instead of an actual date type? I'd think an actual date type would make it easier to query against and work with, though I haven't thought about it too hard.

I did some quick research, which you should absolutely double-check me on, and it looks like the general recommendation here is to use a `date` column with the day component pinned to `1`.

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V27__Create_Enrollmenttable.sql
Comment (karlmdavis) [reply apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V27__Create_Enrollmenttable.sql:137]: Dunno! Depends on _why_ it's null.

File: apps/bfd-pipeline/bfd-pipeline-rif-load/src/test/java/gov/cms/bfd/pipeline/rif/load/RifLoaderIT.java
Comment (karlmdavis) [inline apps/bfd-pipeline/bfd-pipeline-rif-load/src/test/java/gov/cms/bfd/pipeline/rif/load/RifLoaderIT.java:361]: I can't quite work out in my head if your logic above will correctly handle adding a month or two of data to the current year correctly...

You don't really have test coverage for that, so I'd suggest adding it. Maybe just as another step in this test case: what happens when one of the fields moves up to having ten months of data?

**Review by karlmdavis — COMMENTED** [review]
Great work, as always! I do still have a few questions, noted inline.

File: apps/bfd-pipeline/bfd-pipeline-rif-load/src/test/java/gov/cms/bfd/pipeline/rif/load/RifLoaderIT.java
Comment (cbrunefearless) [reply apps/bfd-pipeline/bfd-pipeline-rif-load/src/test/java/gov/cms/bfd/pipeline/rif/load/RifLoaderIT.java:969]: added it to the original 12 months

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V27__Create_Enrollmenttable.sql
Comment (jzulim) [reply apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V27__Create_Enrollmenttable.sql]: From our data providers: As of now, NULL means a beneficiary is accredited but not yet enrolled. We continue to get additional clarity on this.

File: apps/bfd-pipeline/bfd-pipeline-rif-load/src/test/java/gov/cms/bfd/pipeline/rif/load/RifLoaderIT.java
Comment (cbrunefearless) [reply apps/bfd-pipeline/bfd-pipeline-rif-load/src/test/java/gov/cms/bfd/pipeline/rif/load/RifLoaderIT.java]: yes and added it in the test

File: apps/bfd-pipeline/bfd-pipeline-rif-load/src/test/java/gov/cms/bfd/pipeline/rif/load/RifLoaderIT.java
Comment (cbrunefearless) [reply apps/bfd-pipeline/bfd-pipeline-rif-load/src/test/java/gov/cms/bfd/pipeline/rif/load/RifLoaderIT.java:362]: talked about if it makes sense to leave it this way with karl

File: apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java
Comment (karlmdavis) [inline apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java:1944]: FWIW, I'd suggest JavaDoc'ing this, as it's not entirely obvious what some of those parameters mean (I can never quite remember what scale means, for instance).

File: apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java
Comment (karlmdavis) [inline apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java:1876]: FWIW, I'd suggest JavaDoc'ing this, as it's not entirely obvious what some of those parameters mean (I can never quite remember what scale means, for instance).

File: apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java
Comment (karlmdavis) [inline apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java:1832]: FWIW, I'd suggest JavaDoc'ing this, as it's not entirely obvious what some of those parameters mean (I can never quite remember what scale means, for instance).

File: apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java
Comment (karlmdavis) [inline apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java:1971]: FWIW, I'd suggest JavaDoc'ing this, as it's not entirely obvious what some of those parameters mean (I can never quite remember what scale means, for instance).

File: apps/bfd-pipeline/bfd-pipeline-rif-load/src/main/java/gov/cms/bfd/pipeline/rif/load/RifLoader.java
Comment (karlmdavis) [inline apps/bfd-pipeline/bfd-pipeline-rif-load/src/main/java/gov/cms/bfd/pipeline/rif/load/RifLoader.java:909]: FWIW, you don't need this `if` -- the for loop just won't run if the list is empty.

File: apps/bfd-pipeline/bfd-pipeline-rif-load/src/main/java/gov/cms/bfd/pipeline/rif/load/RifLoader.java
Comment (karlmdavis) [inline apps/bfd-pipeline/bfd-pipeline-rif-load/src/main/java/gov/cms/bfd/pipeline/rif/load/RifLoader.java:900]: I'm tired and can't quite think my way through it, but I think this might open us up to a race condition if we ever got a RIF file that listed the same bene multiople times? CSV records are processed in parallel, so there's a chance that you have two threads working on and trying to update the same bene record at the same time.

That'd be weird and bad, but perhaps worth thinking about and at least adding a "`TODO enforce RIF invariant elsewhere: no repeats of same record/PK in same RIF file allowed`".

File: apps/bfd-pipeline/bfd-pipeline-rif-load/src/test/java/gov/cms/bfd/pipeline/rif/load/RifLoaderIT.java
Comment (karlmdavis) [reply apps/bfd-pipeline/bfd-pipeline-rif-load/src/test/java/gov/cms/bfd/pipeline/rif/load/RifLoaderIT.java:362]: I think I said something like, "if I were you and could reason my way to having a lot of confidence in the above logic, I _might_ skip adding the test -- otherwise, I'd double-check myself by adding a test case for it: what happens when we go from having just monthly data for 2020-01 to also having the data for 2020-02?"

**Review by karlmdavis — CHANGES_REQUESTED** [review]
This remains excellent work, Chris!

I did note one possible large problem: I think you need to remove the schema migration from here, lest you have merge conflicts after the first PR. Of course, that will cause the build to fail completely, unless you rebase this PR off of the branch for the other one. :shrug: Or just wait to update this one after that one gets merged.

I also noted one other _possibly_ important thing: depending on your comfort level, you may want to add some more test coverage.

Beyond that, I left a few other minor suggestions.

File: apps/bfd-pipeline/bfd-pipeline-rif-load/src/main/java/gov/cms/bfd/pipeline/rif/load/RifLoader.java
Comment (cbrunefearless) [reply apps/bfd-pipeline/bfd-pipeline-rif-load/src/main/java/gov/cms/bfd/pipeline/rif/load/RifLoader.java:909]: Changed

File: apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/MappingSpec.java
Comment (cbrunefearless) [reply apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/MappingSpec.java:247]: Changed to parentbeneficiary

File: apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java
Comment (cbrunefearless) [reply apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java:1832]: added

File: apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java
Comment (cbrunefearless) [reply apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java:1944]: added

File: apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/MappingSpec.java
Comment (cbrunefearless) [reply apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/MappingSpec.java:265]: changed

File: apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java
Comment (cbrunefearless) [reply apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java:1971]: added

File: apps/bfd-pipeline/bfd-pipeline-rif-load/src/main/java/gov/cms/bfd/pipeline/rif/load/RifLoader.java
Comment (cbrunefearless) [reply apps/bfd-pipeline/bfd-pipeline-rif-load/src/main/java/gov/cms/bfd/pipeline/rif/load/RifLoader.java:900]: added to do

File: apps/bfd-pipeline/bfd-pipeline-rif-load/src/test/java/gov/cms/bfd/pipeline/rif/load/RifLoaderIT.java
Comment (cbrunefearless) [reply apps/bfd-pipeline/bfd-pipeline-rif-load/src/test/java/gov/cms/bfd/pipeline/rif/load/RifLoaderIT.java:362]: added an extra test for august to sept and how the file in sept could update fields in august.

File: apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java
Comment (karlmdavis) [inline apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java:1842]: Would you mind applying this suggestion?
```suggestion
   * @param columnLength specifies the column length {@link Optional<Integer>}, for numeric types this represents the total number of digits that can be stored
   * @param columnScale specifies the column scale {@link Optional<Integer>}, for numeric types this represents how many of the total digits (see `columnLength`) are to the right of the decimal point
```

File: apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java
Comment (karlmdavis) [inline apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java:1897]: Would you mind applying this suggestion?
```suggestion
   * @param columnLength specifies the column length {@link Optional<Integer>}, for numeric types this represents the total number of digits that can be stored
   * @param columnScale specifies the column scale {@link Optional<Integer>}, for numeric types this represents how many of the total digits (see `columnLength`) are to the right of the decimal point
```

File: apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java
Comment (karlmdavis) [inline apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java:1973]: Would you mind applying this suggestion?
```suggestion
   * @param columnLength specifies the column length {@link Optional<Integer>}, for numeric types this represents the total number of digits that can be stored
   * @param columnScale specifies the column scale {@link Optional<Integer>}, for numeric types this represents how many of the total digits (see `columnLength`) are to the right of the decimal point
```

File: apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java
Comment (karlmdavis) [inline apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java:2008]: Would you mind applying this suggestion?
```suggestion
   * @param columnLength specifies the column length {@link Optional<Integer>}, for numeric types this represents the total number of digits that can be stored
   * @param columnScale specifies the column scale {@link Optional<Integer>}, for numeric types this represents how many of the total digits (see `columnLength`) are to the right of the decimal point
```

File: apps/bfd-pipeline/bfd-pipeline-rif-load/src/test/java/gov/cms/bfd/pipeline/rif/load/RifLoaderIT.java
Comment (karlmdavis) [inline apps/bfd-pipeline/bfd-pipeline-rif-load/src/test/java/gov/cms/bfd/pipeline/rif/load/RifLoaderIT.java:688]: What is this assertion doing?

File: apps/bfd-pipeline/bfd-pipeline-rif-load/src/test/java/gov/cms/bfd/pipeline/rif/load/RifLoaderIT.java
Comment (karlmdavis) [inline apps/bfd-pipeline/bfd-pipeline-rif-load/src/test/java/gov/cms/bfd/pipeline/rif/load/RifLoaderIT.java:385]: Great! Glad this passes, too!

**Review by karlmdavis — APPROVED** [review]
LGTM (where the "G" stands for "great"!)

I did leave a couple suggestions and one question, but none of those are blockers (and this'll need re-approval after it gets rebased, anyways).

File: apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java
Comment (cbrunefearless) [reply apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsProcessor.java:545]: yea it did

File: apps/bfd-pipeline/bfd-pipeline-rif-load/src/test/java/gov/cms/bfd/pipeline/rif/load/RifLoaderIT.java
Comment (cbrunefearless) [reply apps/bfd-pipeline/bfd-pipeline-rif-load/src/test/java/gov/cms/bfd/pipeline/rif/load/RifLoaderIT.java:688]: Crap forgot to remove this..

**Review by jzulim — CHANGES_REQUESTED** [review]
As Karl said, Looks GREAT. Thank you for the additional comments in the tests, helped me. This was my first exposure to JavaPoet (more reading to do) but following prior art I was able to step through the mapping code. Built locally with Postgres DB and passing IT's. 

Made one suggestion for a copy/paste typo.

**Review by jzulim — APPROVED** [review]
Ship it

**Review by karlmdavis — APPROVED** [review]
:guitar: Rockin'.

What a slog! Great work, Chris!

**Pull request merged**
